### PR TITLE
Add API key to gRPC server and client

### PR DESCRIPTION
**Pull request by plameniv**

### CODE REVIEW

File: packages/grpc/src/internal/client.ts [hunk 1]
```diff
@@ -86,7 +88,7 @@ export function createGrpcClient<T>(
 
         const metaCallback = (_params: any, callback: (arg0: null, arg1: Metadata) => void) => {
             const meta = new Metadata();
-            meta.add("custom-auth-header", "token");
+            meta.add("custom-auth-header", apiKey || DEFAULT_API_KEY);
```
Comment (YashGajera) [inline packages/grpc/src/internal/client.ts]: I think we can skip adding the header if there is no API Key provided. Also header name should be "authorization"
```suggestion
if(apiKey) {
            meta.add("authorization", apiKey);
            }
```

File: packages/grpc/src/internal/client.ts [hunk 2]
```diff
@@ -75,7 +76,8 @@ class ClientBase implements IRequireInitialization, IDisposable {
 
 export function createGrpcClient<T>(
     config: IGrpcClientConfiguration & IGrpcConfiguration,
-    certPath?: string
+    certPath?: string,
+    apiKey?: string
```
Comment (YashGajera) [inline packages/grpc/src/internal/client.ts]: May be we can turn this into object while this version is still in beta.
option?: {
```suggestion
options?:  {  certPath?: string,
    apiKey?: string
    }
```

File: packages/grpc/src/internal/client.ts [hunk 3]
```diff
@@ -41,6 +41,7 @@ enum GrpcMetricResult {
     Success = "success",
     Error = "error",
 }
+const DEFAULT_API_KEY = "token";
```
Comment (YashGajera) [inline packages/grpc/src/internal/client.ts]: I don't think we need this

File: packages/grpc/src/internal/GrpcInputSource.ts [hunk 4]
```diff
@@ -239,4 +257,12 @@ export class GrpcInputSource implements IInputSource, IRequireInitialization {
             message: error.toString(),
         };
     }
+
+    private isApiKeyValid(meta: Metadata) {
+        const headerValue = meta.get("custom-auth-header");
```
Comment (YashGajera) [inline packages/grpc/src/internal/GrpcInputSource.ts]:
```suggestion
        const headerValue = meta.get("authorization");
```

File: packages/grpc/src/internal/GrpcInputSource.ts [hunk 5]
```diff
@@ -239,4 +257,12 @@ export class GrpcInputSource implements IInputSource, IRequireInitialization {
...
+        if (!headerValue || headerValue.length < 1 || headerValue[0].toString() !== this.apiKey) {
```
Comment (YashGajera) [inline packages/grpc/src/internal/GrpcInputSource.ts:263]:
```suggestion
        if (headerValue?.[0]?.toString() !== this.apiKey) {
```

File: packages/grpc/src/internal/GrpcInputSource.ts [hunk 6]
```diff
@@ -239,4 +257,12 @@ export class GrpcInputSource implements IInputSource, IRequireInitialization {
...
+            return false;
+        }
+        return true;
```
Comment (YashGajera) [inline packages/grpc/src/internal/GrpcInputSource.ts]:
```suggestion
        return headerValue?.[0]?.toString() === this.apiKey;
```

Comment (plameniv) [reply packages/grpc/src/internal/GrpcInputSource.ts]: The suggested correction does not handle an empty array for the headerValue. Do you think that's a unlikely case?

Comment (YashGajera) [reply packages/grpc/src/internal/GrpcInputSource.ts]: It should handle empty array.
Screenshot 2024-05-29 at 2 17 14 PM

File: packages/grpc/src/internal/client.ts (hunk 1 above)
Comment (plameniv) [reply packages/grpc/src/internal/client.ts]: Added the if check a bit earlier

File: packages/grpc/src/internal/client.ts (hunk 2 above)
Comment (plameniv) [reply packages/grpc/src/internal/client.ts]: Created a separate interfaces for server and client

**Review by YashGajera — APPROVED** [review]

File: packages/grpc/src/index.ts [hunk 7]
```diff
@@ -102,7 +112,7 @@ export function grpcMsg(operation: IGrpcServiceMethod, request: any): IMessage {
 
 export function grpcClient<T>(
     configuration: IGrpcClientConfiguration & IGrpcConfiguration,
-    certPath?: string
+    options?: IGrpcClientOptions
```
Comment (cross311) [inline packages/grpc/src/index.ts]: This change is a breaking change.  Can we do it in a non breaking way?  Allow cert to still be passed in and add options and Che k both.

**Review by cross311 — COMMENTED** [review]
Either update version to be a breaking change version or just allow support for old and new way.

**Review by talam — APPROVED** [review]

**Pull request merged**